core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

```diff
@@ -572,28 +573,49 @@ public RequestLogBuilder logBuilder() {
         return log;
     }
 
-    TimeoutScheduler responseTimeoutScheduler() {
-        return responseTimeoutScheduler;
+    CancellationScheduler responseCancellationScheduler() {
+        return responseCancellationScheduler;
+    }
+
+    @Override
+    public void cancel(Throwable cause) {
+        responseCancellationScheduler.finishNow(cause);
+    }
+
+    @Override
+    public void cancel() {
+        cancel(RequestCancellationException.get());
     }
 
     @Override
     public void timeoutNow() {
-        responseTimeoutScheduler.timeoutNow();
+        cancel(ResponseTimeoutException.get());
+    }
+
+    @Nullable
+    @Override
+    public Throwable cancellationCause() {
+        return responseCancellationScheduler.cause();
+    }
+
+    @Override
+    public boolean isCancelled() {
+        return cancellationCause() != null;
     }
 
     @Override
     public boolean isTimedOut() {
-        return responseTimeoutScheduler.isTimedOut();
+        return cancellationCause() instanceof TimeoutException;
     }
 
     @Override
     public CompletableFuture<Void> whenResponseTimingOut() {
```

Comment:
Don't we need to deprecate this?
If then, we also need to migrate the code where `whenResponseTimingOut()` is used.

Reply:
Yes I was thinking we may introduce `whenResponseCancelling/Cancelled` and deprecate that. If ok I'll look to migrate the code.

Reply:
`whenRequestTimingOut()` and `whenRequestTimingOut()` are practically same things. The former is for `ServiceRequestContext` and the latter is for `ClientRequestContext`. We can consolidate them all into `whenCancelling()` and `whenCancelled()`, and do the same deprecation for `whenResponse*()`. What do you think?

Reply:
> `whenRequestTimingOut()` and `whenRequestTimingOut()` are practically same things. The former is for `ServiceRequestContext` and the latter is for `ClientRequestContext`. We can consolidate them all into `whenCancelling()` and `whenCancelled()`, and do the same deprecation for `whenResponse*()`. What do you think?

See my comment below - https://github.com/line/armeria/pull/3080#issuecomment-707566263